### PR TITLE
Istio Status Conditions

### PR DESCRIPTION
**Pull request by videlov**
**Description**

Changes proposed in this pull request:

- Istio Status Conditions

**Related issue(s)**

https://github.com/kyma-project/istio/issues/468

### CODE REVIEW

**Review by barchw — CHANGES_REQUESTED** [review]

File: docs/user/04-00-istio-custom-resource.md [hunk 1]
```diff
@@ -42,4 +42,10 @@ This table lists all the possible parameters of the given resource together with
...
 |----------------------|--------|------------------------------------------------------------------------------------------------------------------------------------|
-| **state** (required) | string | Signifies the current state of **CustomObject**. Its value can be either `Ready`, `Processing`, `Error`, `Warning`, or `Deleting`. |
+| **state** (required) | string | Signifies the current state of the CR. Its value can be either `Ready`, `Processing`, `Error`, `Warning`, or `Deleting`. |
+| **conditions**  | \[\]object | Current conditions state of the CR. |
+| **conditions.&#x200b;lastTransitionTime**  | string | Defines the date of the last condition status change. |
+| **conditions.&#x200b;message**  | string | Provides more details about the condition status change. |
+| **conditions.&#x200b;reason**  | string | Defines the reason for the condition status change. |
+| **conditions.&#x200b;status** (required) | string | Status of the condition. The value is either `True`, `False`, or `Unknown`. |
+| **conditions.&#x200b;type**  | string | Short description of the condition. |
```
Comment (nataliasitko) [inline docs/user/04-00-istio-custom-resource.md:51]:
```suggestion
| **conditions.&#x200b;type**  | string | Provides a short description of the condition. |
```

File: docs/user/04-00-istio-custom-resource.md [hunk 2]
```diff
@@ -42,4 +42,10 @@ This table lists all the possible parameters of the given resource together with
 
 | Parameter            | Type   | Description                                                                                                                        |
 |----------------------|--------|------------------------------------------------------------------------------------------------------------------------------------|
-| **state** (required) | string | Signifies the current state of **CustomObject**. Its value can be either `Ready`, `Processing`, `Error`, `Warning`, or `Deleting`. |
+| **state** (required) | string | Signifies the current state of the CR. Its value can be either `Ready`, `Processing`, `Error`, `Warning`, or `Deleting`. |
+| **conditions**  | \[\]object | Current conditions state of the CR. |
+| **conditions.&#x200b;lastTransitionTime**  | string | Defines the date of the last condition status change. |
+| **conditions.&#x200b;message**  | string | Provides more details about the condition status change. |
+| **conditions.&#x200b;reason**  | string | Defines the reason for the condition status change. |
+| **conditions.&#x200b;status** (required) | string | Status of the condition. The value is either `True`, `False`, or `Unknown`. |
```
Comment (nataliasitko) [inline docs/user/04-00-istio-custom-resource.md:50]:
```suggestion
| **conditions.&#x200b;status** (required) | string | Represents the status of the condition. The value is either `True`, `False`, or `Unknown`. |
```

File: docs/user/04-00-istio-custom-resource.md [hunk 3]
```diff
@@ -42,4 +42,10 @@ This table lists all the possible parameters of the given resource together with
 
 | Parameter            | Type   | Description                                                                                                                        |
 |----------------------|--------|------------------------------------------------------------------------------------------------------------------------------------|
-| **state** (required) | string | Signifies the current state of **CustomObject**. Its value can be either `Ready`, `Processing`, `Error`, `Warning`, or `Deleting`. |
+| **state** (required) | string | Signifies the current state of the CR. Its value can be either `Ready`, `Processing`, `Error`, `Warning`, or `Deleting`. |
+| **conditions**  | \[\]object | Current conditions state of the CR. |
```
Comment (nataliasitko) [inline docs/user/04-00-istio-custom-resource.md:46]:
```suggestion
| **conditions**  | \[\]object | Represents the current state of the CR's conditions. |
```

File: api/v1alpha1/istio_types.go [hunk 4]
```diff
@@ -93,3 +185,29 @@ func (i *Istio) ComponentName() string {
...
+			customMessage = condition.Message
+		}
+		return &metav1.Condition{
+			Type:    string(condition.Type),
+			Status:  condition.Status,
+			Reason:  string(reason),
+			Message: customMessage,
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (videlov) [reply api/v1alpha1/istio_types.go:202]: will add error reporting in logs in `status` package where is the only place the funcs `ConditionFromReason` and `HasReadyCondition` are actually used. Somehow do not think api package should do the logging. Also did unexpose package map variable.

File: api/v1alpha1/istio_types.go [hunk 5]
```diff
@@ -93,3 +185,29 @@ func (i *Istio) ComponentName() string {
 func (i *Istio) HasFinalizer() bool {
 	return len(i.Finalizers) > 0
 }
+
+func ConditionFromReason(reason ConditionReason, customMessage string) *metav1.Condition {
+	condition, found := ConditionReasons[reason]
```
[the remark this answers is not in the document]
Comment (videlov) [reply api/v1alpha1/istio_types.go]: We agreed to keep it for now and refactor later if needed.

File: internal/status/status.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	if len(err.ConditionReasons()) > 0 {
+		conditionReasons = append(conditionReasons, err.ConditionReasons()...)
+	}
+	if !operatorv1alpha1.HasReadyCondition(conditionReasons) {
+		conditionReasons = append(conditionReasons, operatorv1alpha1.NewReasonWithMessage(operatorv1alpha1.ConditionReasonReconcileFailed))
+	}
+	for _, reason := range conditionReasons {
+		d.updateCondition(istioCR, reason.Reason, reason.Message)
```
[the remark this answers is not in the document]
Comment (videlov) [reply internal/status/status.go]: As discussed, I did separate status package funcs for updating status and conditions.

[the remark this answers is not in the document]
Comment (videlov) [reply internal/status/status.go]: We decided to re-discuss this later, created a follow-up task for it.

File: internal/reconciliations/istio/installation.go [hunk 7]
```diff
@@ -194,16 +194,16 @@ func addInstallationFinalizer(ctx context.Context, apiClient client.Client, isti
...
 		}
-		if controllerutil.RemoveFinalizer(istioCR, installationFinalizer) {
-			currentCR.Finalizers = istioCR.Finalizers
-			if err := apiClient.Update(ctx, &currentCR); err != nil {
+		if controllerutil.RemoveFinalizer(&finalizerCR, installationFinalizer) {
+			if err := apiClient.Update(ctx, &finalizerCR); err != nil {
 				return err
 			}
 		}
+		controllerutil.RemoveFinalizer(istioCR, installationFinalizer)
```
Comment (barchw) [inline internal/reconciliations/istio/installation.go:206]: Is ignoring of the return value intentional?

Comment (videlov) [reply internal/reconciliations/istio/installation.go:206]: yes, but will refactor it.

**Review by barchw — APPROVED** [review]

**Review by nataliasitko — APPROVED** [review]

**Pull request merged**